langstream-api-gateway/src/main/java/ai/langstream/apigateway/gateways/LRUTopicProducerCache.java

```diff
@@ -0,0 +1,118 @@
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalNotification;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class LRUTopicProducerCache implements TopicProducerCache {
+
+    private static class SharedTopicProducer implements TopicProducer {
+        private final TopicProducer producer;
+        private volatile int referenceCount;
+        private volatile boolean cached = true;
+
+        public SharedTopicProducer(TopicProducer producer) {
+            this.producer = producer;
+        }
+
+        public void acquire() {
+            synchronized (this) {
+                referenceCount++;
+            }
+        }
+
+        public void removedFromCache() {
+            synchronized (this) {
+                cached = false;
+                if (referenceCount == 0) {
+                    producer.close();
+                }
+            }
+        }
+
+        @Override
+        public void start() {
+            producer.start();
+        }
+
+        @Override
+        public void close() {
+            synchronized (this) {
+                referenceCount--;
+                if (referenceCount == 0 && !cached) {
+                    producer.close();
+                }
+            }
+        }
+
+        @Override
+        public CompletableFuture<?> write(Record record) {
+            return producer.write(record);
+        }
+
+        @Override
+        public Object getNativeProducer() {
+            return producer.getNativeProducer();
+        }
+
+        @Override
+        public Object getInfo() {
+            return producer.getInfo();
+        }
+
+        @Override
+        public long getTotalIn() {
+            return 0;
+        }
+    }
+
+    final Cache<Key, SharedTopicProducer> cache;
+
+    public LRUTopicProducerCache(int size) {
+        this.cache =
+                CacheBuilder.newBuilder()
+                        .maximumSize(size)
+                        .expireAfterWrite(10, TimeUnit.MINUTES)
+                        .expireAfterAccess(10, TimeUnit.MINUTES)
+                        .removalListener(
+                                (RemovalNotification<Key, SharedTopicProducer> notification) -> {
+                                    SharedTopicProducer resource = notification.getValue();
+                                    resource.removedFromCache();
+                                })
+                        .build();
+    }
+
+    @Override
+    public synchronized TopicProducer getOrCreate(
+            TopicProducerCache.Key key, Supplier<TopicProducer> topicProducerSupplier) {
+        try {
+            final SharedTopicProducer sharedTopicProducer =
+                    cache.get(key, () -> new SharedTopicProducer(topicProducerSupplier.get()));
+            sharedTopicProducer.acquire();
```

Comment:
this is a common pitfall

the "acquire" should be performed in the constructor, because between this line and the previous line it is possible that the producer has been released if the count is zero (t is an edge case, but it usually happens).

The tricky thing is that you have to call acquire if there is already an object.

Reply:
Also the producer must be started only once, so you have to start it when you have created it
and the cache should not return the producer until it is fully initialised (started) with success
and if the start operation fails it must not be added to the cache

Reply:
the producer is started in during the load in the cache and every producer in the cache is already started.

Reply:
the acquire must take into condiseration the edge case you mentioned, right. I did it in another manner. When you call acquire, if the producer has been actually closed (removed from cache for example), we retry and that will trigger the constructor to be called